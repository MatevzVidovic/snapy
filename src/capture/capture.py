from __future__ import annotations

import sys
import time
from collections.abc import Callable
from dataclasses import dataclass
from functools import wraps
from pathlib import Path
from typing import Any, ParamSpec, TypeVar
from unittest.mock import MagicMock

import dill

"""
- capture snapshots under `captures/`
- bucket by module + qualname, name files by ms timestamp
- persist args/kwargs/result via dill; ignore capture errors in prod
"""


@dataclass
class CapturePayload:
    args: tuple[Any, ...]
    kwargs: dict[str, Any]
    result: Any


class CaptureHandler:
    @staticmethod
    def get_func_path_id(func: Callable[..., Any]) -> Path:
        """
        - purpose: stable folder ID of a given fn in the form of a path
        - how: use module name (e.g. src.capture.capture), join qualname (RealOps.plus)
        - example return: Path("src/capture/capture/RealOps.plus")
        """
        # - use declared module name (src.capture.capture)
        module_name = func.__module__
        if module_name == "__main__":
            # When fn is in `__main__`, .__module__ is __main__ instead of src.capture.capture.
            # - salvage this using import spec or file path
            module = sys.modules[module_name]
            spec = getattr(module, "__spec__", None)
            if spec and spec.name:
                module_name = spec.name
            else:
                file = getattr(module, "__file__", None)
                if file:
                    module_path = Path(file).resolve().with_suffix("").relative_to(Path.cwd())
                    module_name = ".".join(module_path.parts)

        # - map dotted module + qualname to nested path
        path_id = Path(*module_name.split(".")) / f"{func.__qualname__}"
        return path_id

    @staticmethod
    def get_target_path(func: Callable[..., Any], base_dir: str | Path = "captures") -> Path:
        """
        - purpose: locate capture folder
        - how: combine base dir with normalized function path
        """
        # - build path root + function path id
        base_path = Path(base_dir)
        target_dir = base_path / CaptureHandler.get_func_path_id(func)
        return target_dir

    @staticmethod
    def record(
        args: tuple[Any, ...], kwargs: dict[str, Any], result: Any, target_path: str | Path
    ) -> None:
        """
        - purpose: use dill to save special-snapshots (args, kwargs, return_value)
        """

        # - ensure capture folder exists
        target_path = Path(target_path)
        target_path.mkdir(parents=True, exist_ok=True)
        # - build payload blob
        ts = int(time.time() * 1000)
        payload = CapturePayload(args=args, kwargs=kwargs, result=result)
        blob_path = target_path / f"{ts}.pkl"
        with blob_path.open("wb") as outf:
            dill.dump(payload, outf)

    @staticmethod
    def get_blob_paths(target_path: str | Path) -> list[Path]:
        """
        - purpose: list stored special-snapshots
        - how: guard missing dirs, glob for `.pkl`
        """

        target_path = Path(target_path)
        if not target_path.exists():
            return []
        blobs = list(target_path.glob("*.pkl"))
        return blobs

    @staticmethod
    def get_blob(blob_path: str | Path) -> CapturePayload:
        """
        - purpose: load one snapshot
        - how: open binary path, dill.load
        """
        blob_path = Path(blob_path)
        with blob_path.open("rb") as inf:
            payload = dill.load(inf)
        assert isinstance(payload, CapturePayload)
        return payload


P = ParamSpec("P")
R = TypeVar("R")


def capture(
    max_captures: float = 2, target_path: str | Path | None = None, in_capture_mode: bool = True
) -> Callable[[Callable[P, R]], Callable[P, R]]:
    """
    - purpose: decorator capturing calls when env flag is set
    - how: wrap func; capture only if env var SNAPY_CAPTURE_ENABLED=1;
      if blob count limit reached, do not record more; delegate recording to handler
    """

    def decorator(func: Callable[P, R]) -> Callable[P, R]:
        @wraps(func)
        def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
            # - execute wrapped function first
            result = func(*args, **kwargs)
            try:
                if in_capture_mode:
                    # - lazily resolve target folder
                    nonlocal target_path
                    target_path = (
                        CaptureHandler.get_target_path(func, "captures")
                        if target_path is None
                        else Path(target_path)
                    )
                    # - enforce capture limit
                    num_of_files = (
                        len(list(target_path.glob("*.pkl"))) if target_path.exists() else 0
                    )
                    if num_of_files < max_captures:
                        # - save current call special-snapshot
                        CaptureHandler.record(args, kwargs, result, target_path)
            except Exception:
                pass  # swallow capture errors for now
            return result

        return wrapper

    return decorator


# ---------- Side-effect capture section ----------
"""
Side-effect capture simply reuses regular arg capture mechanism - ctrl+F test_do_ops_DI_with_protocol_mock_snap in test_syrupy.py for how that is done and see the explanation.

Below are simply sugar-fns that make that use a lot cleaner.
"""


def side_effect_lookup(
    args: tuple[Any, ...],
    kwargs: dict[str, Any],
    target_path: str | Path,
    in_test_mode: bool = True,
) -> tuple[Any, bool]:
    """
    - return: (captured_return_val, was_found)
    - use: when making dependency injection mocks, we use @capture to store side-effect return val and fn args.
    We then use this function to load them up:
    this way we do not have to manually create mock return vals and fn args.
    - purpose: reuse stored outputs by arg match
    - how:  - scan existing blobs one by one,
            - if args+kwargs match, return cached result
            - if none matched, signal miss - caller will probably run the actual fn and capture it
    """
    blob_paths = CaptureHandler.get_blob_paths(target_path)
    # - look for payload matching provided args/kwargs
    existing_capture = None
    for bp in blob_paths:
        entry: CapturePayload = CaptureHandler.get_blob(bp)
        if entry.args == args and entry.kwargs == kwargs:
            existing_capture = entry
            break

    if existing_capture is not None:
        # - feed back cached result, was_found=True
        return existing_capture.result, True

    if in_test_mode:
        # - in test mode, missing capture is an error
        raise ValueError("No matching capture found and SIDE_EFFECT_CAPTURE is not set")
    return None, False


def side_effect_target_path(
    test_fn: Callable[..., Any],
    test_case_name: str,
    side_effect_mock: Callable[..., Any],
    storage_base_path: Path = Path("captures", "side_effect_capture"),
) -> Path:
    """
    - purpose: locate side-effect capture folder for test case
    - how: compose captures/side_effect_capture with function + mock + test name
    """
    return (
        storage_base_path
        / CaptureHandler.get_func_path_id(test_fn)
        / test_case_name
        / side_effect_mock.__qualname__
    )


def assert_side_effect_calls(
    test_fn: Callable[..., Any],
    test_case_name: str,
    side_effect_mock: Callable[..., Any],
    magic_mock: MagicMock,
) -> None:
    """
    - purpose: replay captured args against provided mock
    - how: load stored blobs and assert mock called with each payload
    """
    target_path = side_effect_target_path(test_fn, test_case_name, side_effect_mock)
    blobs = CaptureHandler.get_blob_paths(target_path)

    for bp in blobs:
        # - assert recorded call signature matches mock usage
        entry: CapturePayload = CaptureHandler.get_blob(bp)
        magic_mock.assert_called_with(*entry.args, **entry.kwargs)
