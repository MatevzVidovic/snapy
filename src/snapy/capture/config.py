"""
Configuration management for snap capture functionality.
"""

import os
import re
from pathlib import Path
from typing import List, Optional, Dict, Any
from dataclasses import dataclass, field


@dataclass
class CaptureConfig:
    """
    Configuration class for snap capture functionality.

    Manages settings from environment variables, .env files, and explicit parameters.
    """

    # Default configuration values
    enabled: bool = field(default_factory=lambda: os.getenv("SNAP_CAPTURE_ENABLED", "true").lower() == "true")
    default_path: str = field(default_factory=lambda: os.getenv("SNAP_CAPTURE_DEFAULT_PATH", "./snap_capture"))
    default_retention: int = field(default_factory=lambda: int(os.getenv("SNAP_CAPTURE_DEFAULT_RETENTION", "2")))
    default_overwrite: bool = field(default_factory=lambda: os.getenv("SNAP_CAPTURE_DEFAULT_OVERWRITE", "false").lower() == "true")

    # Filtering configuration
    ignore_modules: List[str] = field(default_factory=lambda: CaptureConfig._parse_env_list("SNAP_CAPTURE_IGNORE_MODULES"))
    ignore_functions: List[str] = field(default_factory=lambda: CaptureConfig._parse_env_list("SNAP_CAPTURE_IGNORE_FUNCTIONS"))
    ignore_args: List[str] = field(default_factory=lambda: CaptureConfig._parse_env_list("SNAP_CAPTURE_IGNORE_ARGS",
                                                                                         ["password", "token", "secret", "key", "auth"]))

    # Production optimization
    production_mode: bool = field(default_factory=lambda: os.getenv("SNAP_CAPTURE_PRODUCTION_MODE", "false").lower() == "true")
    minimal_capture: bool = field(default_factory=lambda: os.getenv("SNAP_CAPTURE_MINIMAL", "false").lower() == "true")

    @staticmethod
    def _parse_env_list(env_var: str, default: Optional[List[str]] = None) -> List[str]:
        """Parse comma-separated environment variable into a list."""
        value = os.getenv(env_var)
        if value:
            return [item.strip() for item in value.split(",") if item.strip()]
        return default or []

    @classmethod
    def from_env_file(cls, env_file_path: Optional[str] = None) -> "CaptureConfig":
        """
        Create configuration from .env file.

        Args:
            env_file_path: Path to .env file. If None, looks for .env in current directory.

        Returns:
            CaptureConfig instance with values loaded from .env file
        """
        if env_file_path is None:
            env_file_path = ".env"

        env_path = Path(env_file_path)
        if env_path.exists():
            cls._load_env_file(env_path)

        return cls()

    @staticmethod
    def _load_env_file(env_path: Path) -> None:
        """Load environment variables from .env file."""
        with open(env_path, 'r') as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith('#') and '=' in line:
                    key, value = line.split('=', 1)
                    key = key.strip()
                    value = value.strip().strip('"').strip("'")
                    os.environ.setdefault(key, value)

    def is_function_ignored(self, function_name: str, module_name: str) -> bool:
        """
        Check if a function should be ignored based on configuration.

        Args:
            function_name: Name of the function
            module_name: Name of the module containing the function

        Returns:
            True if the function should be ignored, False otherwise
        """
        # Check module patterns
        for pattern in self.ignore_modules:
            if self._match_pattern(pattern, module_name):
                return True

        # Check function patterns
        for pattern in self.ignore_functions:
            if self._match_pattern(pattern, function_name):
                return True

        return False

    def is_arg_ignored(self, arg_name: str) -> bool:
        """
        Check if an argument should be ignored based on configuration.

        Args:
            arg_name: Name of the argument

        Returns:
            True if the argument should be ignored, False otherwise
        """
        for pattern in self.ignore_args:
            if self._match_pattern(pattern, arg_name):
                return True
        return False

    @staticmethod
    def _match_pattern(pattern: str, text: str) -> bool:
        """
        Match a pattern against text using wildcards.

        Args:
            pattern: Pattern with possible wildcards (*)
            text: Text to match against

        Returns:
            True if pattern matches, False otherwise
        """
        # Convert wildcard pattern to regex
        regex_pattern = pattern.replace("*", ".*")
        regex_pattern = f"^{regex_pattern}$"

        try:
            return bool(re.match(regex_pattern, text, re.IGNORECASE))
        except re.error:
            # If regex is invalid, fall back to simple string matching
            return pattern.lower() in text.lower()

    def get_capture_path(self, custom_path: Optional[str] = None) -> Path:
        """
        Get the capture path, with optional override.

        Args:
            custom_path: Custom path to use instead of default

        Returns:
            Path object for capture storage
        """
        path_str = custom_path or self.default_path
        path = Path(path_str)

        # Create directory if it doesn't exist
        path.mkdir(parents=True, exist_ok=True)

        return path

    def should_capture(self, function_name: str, module_name: str) -> bool:
        """
        Determine if capturing should be performed for a given function.

        Args:
            function_name: Name of the function
            module_name: Name of the module

        Returns:
            True if capturing should be performed, False otherwise
        """
        if not self.enabled:
            return False

        if self.is_function_ignored(function_name, module_name):
            return False

        return True

    def to_dict(self) -> Dict[str, Any]:
        """
        Convert configuration to dictionary.

        Returns:
            Dictionary representation of configuration
        """
        return {
            "enabled": self.enabled,
            "default_path": self.default_path,
            "default_retention": self.default_retention,
            "default_overwrite": self.default_overwrite,
            "ignore_modules": self.ignore_modules,
            "ignore_functions": self.ignore_functions,
            "ignore_args": self.ignore_args,
            "production_mode": self.production_mode,
            "minimal_capture": self.minimal_capture
        }


# Global configuration instance
_global_config: Optional[CaptureConfig] = None


def get_global_config() -> CaptureConfig:
    """
    Get the global configuration instance.

    Returns:
        Global CaptureConfig instance
    """
    global _global_config
    if _global_config is None:
        _global_config = CaptureConfig.from_env_file()
    return _global_config


def set_global_config(config: CaptureConfig) -> None:
    """
    Set the global configuration instance.

    Args:
        config: CaptureConfig instance to set as global
    """
    global _global_config
    _global_config = config