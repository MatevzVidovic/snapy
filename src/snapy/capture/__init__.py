"""
Snapy Capture - Automatic Function Argument Capture for Replay Testing

A decorator-based system for capturing function arguments and enabling easy
replay in snapshot tests without manually creating test arguments.
"""

from .config import CaptureConfig
from .storage import CaptureStorage
from .filters import ArgumentFilter
from .capture import capture_args
from .loader import CaptureLoader, load_capture, load_capture_dict

__version__ = "0.1.0"
__all__ = [
    "capture_args",
    "CaptureConfig",
    "CaptureStorage",
    "ArgumentFilter",
    "CaptureLoader",
    "load_capture",
    "load_capture_dict"
]