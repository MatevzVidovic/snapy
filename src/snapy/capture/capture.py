"""
Main capture decorator for function argument capture.
"""

import functools
import inspect
import asyncio
from collections import OrderedDict
from typing import Callable, Optional, List, Any, TypeVar, Union, Dict
from .config import CaptureConfig, get_global_config
from .storage import CaptureStorage
from .filters import ArgumentFilter

F = TypeVar('F', bound=Callable[..., Any])


def capture_args(
    path: Optional[str] = None,
    retention: Optional[int] = None,
    overwrite: Optional[bool] = None,
    ignore_args: Optional[List[str]] = None,
    config: Optional[CaptureConfig] = None,
    enabled: Optional[bool] = None
) -> Callable[[F], F]:
    """
    Decorator that captures function arguments for replay in tests.

    Args:
        path: Custom storage path (overrides config default)
        retention: Number of captures to retain (overrides config default)
        overwrite: Whether to overwrite existing captures (overrides config default)
        ignore_args: Additional argument names to ignore (extends config)
        config: Custom CaptureConfig instance (overrides global config)
        enabled: Override global enable/disable setting

    Returns:
        Decorated function that captures arguments when called

    Example:
        @capture_args(path="./my_captures", retention=3)
        def process_data(user_id, data, password=None):
            return {"processed": data, "user": user_id}

        # When called, arguments are automatically captured
        result = process_data("user123", {"key": "value"}, password="secret")
    """

    def decorator(func: F) -> F:
        # Get configuration
        effective_config = config or get_global_config()

        # Override config values if provided
        if ignore_args:
            # Create a copy of config with additional ignore patterns
            import copy
            effective_config = copy.deepcopy(effective_config)
            effective_config.ignore_args.extend(ignore_args)

        # Get function metadata
        function_name = func.__name__
        module_name = func.__module__ or ""

        # Check if this function should be captured
        capture_enabled = enabled if enabled is not None else effective_config.enabled
        if not capture_enabled or effective_config.is_function_ignored(function_name, module_name):
            # Return original function unchanged if disabled or ignored
            return func

        # Initialize storage and filter
        storage = CaptureStorage(effective_config.get_capture_path(path))
        arg_filter = ArgumentFilter(effective_config)

        # Handle both sync and async functions
        if asyncio.iscoroutinefunction(func):
            @functools.wraps(func)
            async def async_wrapper(*args, **kwargs):
                return await _capture_and_call_async(
                    func, args, kwargs, function_name, module_name,
                    storage, arg_filter, effective_config, retention, overwrite
                )
            return async_wrapper
        else:
            @functools.wraps(func)
            def sync_wrapper(*args, **kwargs):
                return _capture_and_call_sync(
                    func, args, kwargs, function_name, module_name,
                    storage, arg_filter, effective_config, retention, overwrite
                )
            return sync_wrapper

    return decorator


def _capture_and_call_sync(
    func: Callable,
    args: tuple,
    kwargs: dict,
    function_name: str,
    module_name: str,
    storage: CaptureStorage,
    arg_filter: ArgumentFilter,
    config: CaptureConfig,
    retention: Optional[int],
    overwrite: Optional[bool]
) -> Any:
    """Handle capture and execution for synchronous functions."""
    try:
        # Capture arguments before function execution
        _capture_arguments(
            func, args, kwargs, function_name, module_name,
            storage, arg_filter, config, retention, overwrite
        )
    except Exception as e:
        # Don't fail the original function if capture fails
        print(f"Warning: Argument capture failed for {function_name}: {e}")

    # Execute original function
    return func(*args, **kwargs)


async def _capture_and_call_async(
    func: Callable,
    args: tuple,
    kwargs: dict,
    function_name: str,
    module_name: str,
    storage: CaptureStorage,
    arg_filter: ArgumentFilter,
    config: CaptureConfig,
    retention: Optional[int],
    overwrite: Optional[bool]
) -> Any:
    """Handle capture and execution for asynchronous functions."""
    try:
        # Capture arguments before function execution
        _capture_arguments(
            func, args, kwargs, function_name, module_name,
            storage, arg_filter, config, retention, overwrite
        )
    except Exception as e:
        # Don't fail the original function if capture fails
        print(f"Warning: Argument capture failed for {function_name}: {e}")

    # Execute original function
    return await func(*args, **kwargs)


def _args_to_named_dict(func: Callable, args: tuple, kwargs: dict) -> OrderedDict[str, Any]:
    """
    Convert positional and keyword arguments to a named dictionary using function signature.

    Args:
        func: The function being called
        args: Positional arguments tuple
        kwargs: Keyword arguments dict

    Returns:
        OrderedDict mapping parameter names to values in signature order
    """
    try:
        # Get function signature
        sig = inspect.signature(func)
        bound_args = sig.bind(*args, **kwargs)
        bound_args.apply_defaults()

        # Create ordered dict preserving parameter order
        named_args = OrderedDict()
        for param_name in sig.parameters.keys():
            if param_name in bound_args.arguments:
                named_args[param_name] = bound_args.arguments[param_name]

        return named_args

    except (TypeError, ValueError) as e:
        # Fallback: create basic mapping if signature binding fails
        print(f"Warning: Could not bind arguments to function signature: {e}")

        # Try to get parameter names from function
        try:
            param_names = list(inspect.signature(func).parameters.keys())
            named_args = OrderedDict()

            # Map positional args to parameter names
            for i, arg_value in enumerate(args):
                if i < len(param_names):
                    named_args[param_names[i]] = arg_value
                else:
                    # Extra positional args
                    named_args[f"*args[{i}]"] = arg_value

            # Add keyword arguments
            for key, value in kwargs.items():
                named_args[key] = value

            return named_args

        except Exception:
            # Ultimate fallback: just use indices for positional args
            named_args = OrderedDict()
            for i, arg_value in enumerate(args):
                named_args[f"arg_{i}"] = arg_value
            for key, value in kwargs.items():
                named_args[key] = value

            return named_args


def _capture_arguments(
    func: Callable,
    args: tuple,
    kwargs: dict,
    function_name: str,
    module_name: str,
    storage: CaptureStorage,
    arg_filter: ArgumentFilter,
    config: CaptureConfig,
    retention: Optional[int],
    overwrite: Optional[bool]
) -> None:
    """Capture function arguments with filtering and storage."""
    # Use config defaults if not specified
    effective_retention = retention if retention is not None else config.default_retention
    effective_overwrite = overwrite if overwrite is not None else config.default_overwrite

    # Convert args to named arguments using function signature
    all_args_dict = _args_to_named_dict(func, args, kwargs)

    # Filter arguments
    if arg_filter.should_capture_minimal():
        # Minimal capture for production
        filtered_args_dict = arg_filter.get_minimal_args_dict(all_args_dict)
    else:
        # Full capture with filtering
        filtered_args_dict = arg_filter.filter_args_dict(all_args_dict)

    # Save capture
    storage.save_capture(
        function_name=function_name,
        module_name=module_name,
        args_dict=filtered_args_dict,
        retention=effective_retention,
        overwrite=effective_overwrite
    )


class CaptureContext:
    """
    Context manager for temporarily enabling/disabling capture for a block of code.

    Example:
        with CaptureContext(enabled=False):
            # No captures will be made in this block
            some_decorated_function()

        with CaptureContext(path="./special_captures"):
            # Captures will use special path
            another_decorated_function()
    """

    def __init__(
        self,
        enabled: Optional[bool] = None,
        path: Optional[str] = None,
        retention: Optional[int] = None,
        config: Optional[CaptureConfig] = None
    ):
        """
        Initialize capture context.

        Args:
            enabled: Override global enabled setting
            path: Override storage path
            retention: Override retention setting
            config: Override global config
        """
        self.enabled = enabled
        self.path = path
        self.retention = retention
        self.config = config
        self._original_config = None

    def __enter__(self):
        """Enter context and apply overrides."""
        global _global_config
        from .config import _global_config

        # Store original config
        self._original_config = _global_config

        # Create modified config if needed
        if any([self.enabled is not None, self.path is not None,
                self.retention is not None, self.config is not None]):
            import copy

            if self.config:
                new_config = copy.deepcopy(self.config)
            else:
                new_config = copy.deepcopy(get_global_config())

            # Apply overrides
            if self.enabled is not None:
                new_config.enabled = self.enabled
            if self.path is not None:
                new_config.default_path = self.path
            if self.retention is not None:
                new_config.default_retention = self.retention

            # Set as global config
            from .config import set_global_config
            set_global_config(new_config)

        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Exit context and restore original config."""
        if self._original_config is not None:
            from .config import set_global_config
            set_global_config(self._original_config)


# Convenience functions
def disable_capture():
    """
    Decorator to disable capture for a specific function.

    Example:
        @disable_capture()
        @capture_args()
        def my_function():
            pass  # This function won't capture arguments
    """
    return capture_args(enabled=False)


def capture_once(path: Optional[str] = None):
    """
    Decorator that captures arguments only once (overwrite=False).

    Args:
        path: Optional custom storage path

    Example:
        @capture_once()
        def expensive_function(large_data):
            pass  # Arguments captured only on first call
    """
    return capture_args(path=path, overwrite=False, retention=1)


def capture_minimal(path: Optional[str] = None):
    """
    Decorator that captures only minimal argument information.

    Args:
        path: Optional custom storage path

    Example:
        @capture_minimal()
        def production_function(sensitive_data):
            pass  # Only captures argument types, not values
    """
    import copy
    config = copy.deepcopy(get_global_config())
    config.minimal_capture = True
    return capture_args(path=path, config=config)