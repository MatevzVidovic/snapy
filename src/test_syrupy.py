
import os
from pathlib import Path
import examples.basics as b

from unittest.mock import Mock
from pytest_mock import MockerFixture, MockType


def test_greeting():
    b.main()

# tests/test_greeting.py
def test_do_ops(snapshot):

    returned = b.do_ops(b.RealOpsOne(), 1, 2)
    assert returned == snapshot


def test_do_ops_with_protocol_mock(mocker, snapshot):
    ops = mocker.create_autospec(b.BasicOps, instance=True, spec_set=True)
    ops.plus.return_value = 3
    # THIS IS CLARLY PROBLEMATIC.
    # The 2 lines below are wrong - that is not at is returned.
    # But since it is never used as an input, 
    # we dont even know that that is what is returned.
    ops.expression.return_value = "expr(1, 2)"
    ops.concatenation.return_value = "concat(1, 2)"

    print_spy = mocker.patch("examples.basics.print")

    b.do_ops(ops, 1, 2)

    ops.plus.assert_called_once_with(1, 2, c=8)
    ops.expression.assert_called_once_with(1, 2)
    ops.concatenation.assert_called_once_with(1, 2)
    assert [call.args for call in print_spy.call_args_list] == snapshot



def test_do_ops_DI_with_protocol_mock(mocker, snapshot):

    real_ops = b.RealOpsOne()

    ops = mocker.create_autospec(b.BasicOps, instance=True, spec_set=True)
    ops.plus.return_value = 3
    ops.expression.side_effect = real_ops.expression
    ops.concatenation.side_effect = real_ops.concatenation

    # print_spy = mocker.patch(real_ops, "concatenation", wraps=real_ops.concatenation)
    # print_spy = mocker.patch("examples.basics.RealOpsOne.concatenation")
    # print_spy = mocker.patch(real_ops, "concatenation")
    print_spy = mocker.patch("examples.basics.RealOpsOne.concatenation", wraps=real_ops.concatenation)


    # print_spy = mocker.patch("examples.basics.print")

    returned = b.do_ops_DI(ops, 1, 2)

    print(f"\nspied: {print_spy.call_args_list}")

    ops.plus.assert_called_once_with(1, 2)
    ops.expression.assert_called_once()
    ops.concatenation.assert_called()

    # assert [call.args for call in print_spy.call_args_list] == snapshot

    assert returned == snapshot


import src.capture.capture as c
cch = c.CaptureHandler
# def test_real_ops_one_concatenation(snapshot):

#     table = cch().load_all(b.RealOpsOne.concatenation)
#     print(f"table: {table}")

#     for _, entry in table.items():
#         result = b.RealOpsOne().concatenation(*entry["args"], **entry["kwargs"])
#         assert result == snapshot

def test_real_ops_one_plus(snapshot):

    table = cch.load_all(b.RealOpsOne.plus)
    print(f"table: {table}")

    for _, entry in table.items():
        # print(b.RealOpsOne().plus(5, 8))
        result = b.RealOpsOne().plus(*entry["args"][1:], **entry["kwargs"])
        # print(f"result: {result}")
        assert result == snapshot





def test_do_ops_DI_with_protocol_mock_snap(mocker: MockerFixture, snapshot):

    """

    - v plus_mock imamo potem fn plus_wrapper ki ima v sebi potem plus().
    
    In če je plus_wrapper že imel take argumente (vidimo v capture load) potem samo returnamo to.

    Če še ni imel, poženemo plus wrapper (na njim je @capture torej se bo zdaj shranilo).

    - V testu imamo pogledamo za dotenv spremenljivko .
    Če je ta false ali none, in ne najdemo matching signaturea v loaded captures, damo Exception.
    """

    table_tests = cch.load_all(b.do_ops)
    for test_case, captures in table_tests.items():
        ops = mocker.create_autospec(b.BasicOps, instance=True, spec_set=True)

        def plus_mock(*args, **kwargs):
            import src.capture.capture as c
            side_effect_target_path = c.side_effect_target_path(test_do_ops_DI_with_protocol_mock_snap, plus_mock, test_case)
            returned, was_found = c.side_effect_lookup(args, kwargs, side_effect_target_path)
            if was_found:
                return returned
            
            @c.capture(max_captures=float("inf"), target_path=side_effect_target_path)
            def plus_wrapper(*args, **kwargs):
                import examples.basics as b
                return b.RealOpsOne.plus(*args, **kwargs)
            return plus_wrapper(*args, **kwargs)


        def expression_mock(*args, **kwargs):
            import src.capture.capture as c
            side_effect_target_path = c.side_effect_target_path(test_do_ops_DI_with_protocol_mock_snap, expression_mock, test_case)
            returned, was_found = c.side_effect_lookup(args, kwargs, side_effect_target_path)
            if was_found:
                return returned
            
            @c.capture(max_captures=float("inf"), target_path=side_effect_target_path)
            def expression_wrapper(*args, **kwargs):
                import examples.basics as b
                return b.RealOpsOne.expression(*args, **kwargs)
            return expression_wrapper(*args, **kwargs)
        

        def concatenation_mock(*args, **kwargs):
            import src.capture.capture as c
            side_effect_target_path = c.side_effect_target_path(test_do_ops_DI_with_protocol_mock_snap, concatenation_mock, test_case)
            returned, was_found = c.side_effect_lookup(args, kwargs, side_effect_target_path)
            if was_found:
                return returned
            
            @c.capture(max_captures=float("inf"), target_path=side_effect_target_path)
            def concatenation_wrapper(*args, **kwargs):
                import examples.basics as b
                return b.RealOpsOne.concatenation(*args, **kwargs)
            return concatenation_wrapper(*args, **kwargs)


    

        ops.plus.side_effect = plus_mock
        ops.expression.side_effect = expression_mock
        ops.concatenation.side_effect = concatenation_mock

        print_spy = mocker.patch("examples.basics.print")

        returned = b.do_ops_DI(*captures["args"], **captures["kwargs"])

        c.assert_side_effect_calls(test_do_ops_DI_with_protocol_mock_snap, concatenation_mock, test_case, ops.plus)
        c.assert_side_effect_calls(test_do_ops_DI_with_protocol_mock_snap, expression_mock, test_case, ops.expression)
        c.assert_side_effect_calls(test_do_ops_DI_with_protocol_mock_snap, concatenation_mock, test_case, ops.concatenation)

        assert [returned] + [call.args for call in print_spy.call_args_list] == snapshot
        
        