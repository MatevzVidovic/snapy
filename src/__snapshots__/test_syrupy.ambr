# serializer version: 1
# name: test_do_ops
  '336.2'
# ---
# name: test_do_ops_DI_with_protocol_mock
  'start336.2'
# ---
# name: test_do_ops_with_protocol_mock
  list([
  ])
# ---
